torch/csrc/jit/codegen/cuda/scheduler/transpose.cpp

```diff
@@ -0,0 +1,637 @@
+  // inputs and outputs themselves are disconnected, so we have to borrow the
+  // entire DAG and use its spanning tree.
+  {
+    auto all_tvs_except1 = ir_utils::allTvsExcept(
+        fusion,
+        {grouped_inputs_outputs[0].begin(), grouped_inputs_outputs[0].end()});
+    SetSelector selector({all_tvs_except1.begin(), all_tvs_except1.end()});
+    MaxRootDomainInfoSpanningTree entire_dag_except1(reference2, &selector);
+    TransformPropagator propagator(reference2);
+    entire_dag_except1.traverse(&propagator);
+  }
+
+  // schedule group 1
+  reference1->reorder({{-2, -1}});
+  // [..., tile2, tile1]
+  pos = reference1->nDims() - 2;
+  reference1->merge(pos);
+  reference1->split(pos, params.vectorize_factor1);
+  reference1->split(pos, warp_size);
+  reference1->split(pos, unroll_factor1, false);
+  reference1->axis(-1)->parallelize(ParallelType::Vectorize);
+  reference1->axis(-2)->parallelize(ParallelType::TIDx);
+  reference1->axis(-3)->parallelize(ParallelType::TIDy);
+  reference1->axis(-4)->parallelize(ParallelType::Unroll);
+  // [..., Unroll, TIDy, TIDx, Vectorize]
+
+  // Propagate transformations, parallelization of the reference1 to the entire
+  // DAG except group 2 and its corresponding cached outputs.
+  {
+    auto all_tvs_except2 =
+        ir_utils::allTvsExcept(fusion, group2_and_cached_inputs);
+    SetSelector selector({all_tvs_except2.begin(), all_tvs_except2.end()});
+    MaxRootDomainInfoSpanningTree entire_dag_except_outputs(
+        reference1, &selector);
+    TransformPropagator propagator(reference1);
+    entire_dag_except_outputs.traverse(&propagator);
+    scheduler_utils::parallelizeAllLike(
+        reference1, all_tvs_except2, {ParallelType::TIDx, ParallelType::TIDy});
+  }
+
+  // vectorize and unroll group 1's output and cached input
+  {
+    std::vector<TensorView*> group1_and_cached_inputs(
+        grouped_inputs_outputs[0].begin(), grouped_inputs_outputs[0].end());
+    for (auto tv : grouped_inputs_outputs[0]) {
+      if (tv->isFusionInput()) {
+        group1_and_cached_inputs.emplace_back(ir_utils::consumerTvsOf(tv)[0]);
+      }
+    }
+    scheduler_utils::parallelizeAllLike(
+        reference2,
+        group1_and_cached_inputs,
+        {ParallelType::Vectorize, ParallelType::Unroll});
+  }
+
+  // parallelize group2
+  scheduler_utils::parallelizeAllLike(
+      reference2,
+      {group2_and_cached_inputs.begin(), group2_and_cached_inputs.end()},
+      {ParallelType::Vectorize,
+       ParallelType::Unroll,
+       ParallelType::TIDx,
+       ParallelType::TIDy});
+
+  // cleanup parallelization from reference1 and reference2 if they are fusion
+  // inputs
+  for (auto tv : {reference1, reference2}) {
+    if (tv->isFusionInput()) {
+      for (auto id : tv->domain()->domain()) {
+        id->parallelize(ParallelType::Serial);
+      }
+    }
+  }
+
+  // Inline
+  InlinePropagator inline_propagator(reference1, 2, ComputeAtMode::BestEffort);
```

Comment:
If we have a long complex dag, wouldn't we want vector dims to be inlined except the smem tiles and vectorized reads/writes? Otherwise we might have relatively high register pressure.

Reply:
Changed to most inlined propagation, but needs https://github.com/csarofeen/pytorch/pull/1868